Examples/Framework/include/ActsExamples/Utilities/tbbWrap.hpp

```diff
@@ -0,0 +1,169 @@
+/// (requires nthreads=1 or -1). This allows the ACTS Examples to be built
+/// without the tbb library (and reduces the dependency on ROOT).
+/// Based on an idea from
+///   https://stackoverflow.com/questions/59736661/how-to-completely-switch-off-threading-in-tbb-code
+
+#ifdef NO_TBB
+namespace tbb {
+struct task_arena {
+  static const int automatic = -1;
+};
+
+template <typename Value>
+struct blocked_range {
+  blocked_range(Value begin_, Value end_) : my_end(end_), my_begin(begin_) {}
+  Value begin() const { return my_begin; }
+  Value end() const { return my_end; }
+
+ private:
+  Value my_end;
+  Value my_begin;
+};
+}  // namespace tbb
+#endif
+
+namespace tbbWrap {
+static bool enableTBB(int nthreads = -99) {
+  static bool setting = false;
+  if (nthreads != -99) {
+#ifdef NO_TBB
+    if (nthreads > 1) {
+      throw std::runtime_error(
+          "tbb is not supported, so can't do multi-threading.");
+    }
+#else
+    bool newSetting = (nthreads != 1);
+    if (newSetting != setting) {
+      if (newSetting) {
+        // std::cout << "Enable TBB" << std::endl;
+        setting = newSetting;
+      } else {
+        // std::cout << "Don't disable TBB, since it is already in use." <<
+        // std::endl;
+      }
+    }
+#endif
+  }
+  // std::cout << "TBB is " << (setting ? "enabled" : "disabled") << std::endl;
+  return setting;
+}
+
+class task_arena {
+#ifndef NO_TBB
+  std::unique_ptr<tbb::task_arena> tbb;
+#endif
+
+ public:
+#ifdef NO_TBB
+  task_arena(int nthreads = tbb::task_arena::automatic, unsigned = 1){
+#else
+  task_arena(int nthreads = tbb::task_arena::automatic, unsigned res = 1) {
+#endif
+      if (enableTBB(nthreads)) {
+#ifndef NO_TBB
+        tbb.reset(new tbb::task_arena(nthreads, res));
+#endif
+}
+}  // namespace tbbWrap
+
+template <typename F>
+void execute(const F& f) {
+#ifndef NO_TBB
+  if (tbb) {
+    tbb->execute(f);
+  } else
+#endif
+  {
+    f();
+  }
+}
+};  // namespace ActsExamples
+
+class parallel_for {
+ public:
+  template <typename R, typename F>
+  parallel_for(const R& r, const F& f) {
+#ifndef NO_TBB
+    if (enableTBB()) {
+      tbb::parallel_for(r, f);
+    } else
+#endif
+    {
+      f(r);
+    }
+  }
+};
+
+class queuing_mutex {
+#ifndef NO_TBB
+  std::unique_ptr<tbb::queuing_mutex> tbb;
+#endif
+
+ public:
+  queuing_mutex() {
+#ifndef NO_TBB
+    if (enableTBB()) {
+      tbb.reset(new tbb::queuing_mutex());
+    }
+#endif
+  }
+
+  class scoped_lock {
+#ifndef NO_TBB
+    std::unique_ptr<tbb::queuing_mutex::scoped_lock> tbb;
+#endif
+
+   public:
+    scoped_lock() {
+#ifndef NO_TBB
+      if (enableTBB()) {
+        tbb.reset(new tbb::queuing_mutex::scoped_lock());
+      }
+#endif
+    }
+
+#ifdef NO_TBB
+    scoped_lock(queuing_mutex&){
+#else
+    scoped_lock(queuing_mutex& m) {
+      if (enableTBB()) {
+        tbb.reset(new tbb::queuing_mutex::scoped_lock(*m.tbb.get()));
+      }
+#endif
+    }
+  };
+};
```

Comment:
I'm very uneasy about having lock classes that do nothing in some configurations.

To me, this raises the more philosophical question of whether we actually want to remove all uses of the TBB library, or just the parallelization part (task_arena and parallel_for). I would argue that the latter is sufficient to achieve your stated goal of simplifying debugging through direct calls and reduced thread spawning.

Also, if you want to stick with the current approach, consider using std::optional instead of std::unique_ptr.

Reply:
Sticking with `tbb::queuing_mutex` without wrapping it would certainly be cleaner. I wasn't sure whether it would work without initialising the rest of the `tbb` library, thread pool and everything.
Even if we can determine that's OK, would it play nice with ROOT where we don't call `ROOT::EnableThreadSafety()`? I really don't want to enable that in the unthreaded case, because that introduces even more confusions.

Reply:
I just checked, and if I run with a single thread and don't do any ROOT I/O with the Generic detector, then ACTS doesn't need to initialise ROOT at all, so no longer forks anything. I don't think the ROOT stuff interferes much, but this is a nice side-effect.